### ISSUE TITLE
Augment TensorPrimitives for post-.NET 8

### ISSUE THREAD

**Issue by stephentoub** [dotnet/runtime#93286]
Regardless of any additional types we may want to add to `System.Numerics.Tensors`, we would like to expand the set of APIs exposed on the `TensorPrimitives` static class in a few ways:
- Additional operations from `Math{F}` that don't currently have representation on `TensorPrimitives`, e.g. `CopySign`, `Reciprocal{Sqrt}{Estimate}`, `Sqrt`, `Ceiling`, `Floor`, `Truncate`, `Log10`, `Log(x, y)` (with y as both span and scalar), `Pow(x, y)` (with y as both span and scalar), `Cbrt`, `IEEERemainder`, `Acos`, `Acosh`, `Cos`, `Asin`, `Asinh`, `Sin`, `Atan`. [This unmerged commit](https://github.com/dotnet/runtime/commit/ada9b18f16ab6c248fe10deedb22404802334309) has a sketch, but it's out-of-date with improvements that have been made to the library since, and all of the operations should be vectorized.
- Additional operations defined in the numerical interfaces that don't currently have representation on `TensorPrimitives`, e.g. `BitwiseAnd`, `BitwiseOr`, `BitwiseXor`, `Exp10`, `Exp10M1`, `Exp2`, `Exp2M1`, `ExpM1`, `Atan2`, `Atan2Pi`, `ILogB`, `Lerp`, `ScaleB`, `Round`, `Log10P1`, `Log2P1`, `LogP1`, `Hypot`, `RootN`, `AcosPi`, `AsinPi`, `AtanPi`, `CosPi`, `SinPi`, `TanPi`
- Additional operations defined in BLAS / LAPACK that don't currently have representation on `TensorPrimitives`
- Additional operations that would enable completely removing the internal `CpuMath` class from ML.NET, e.g.  `Add` (with indices), `AddScale` (with indices), `DotProductSparse`, `MatrixTimesSource`, `ScaleAdd` improvement via `AddMultiply` or `MultipleAdd` overloads, `SdcaL1UpdateDense`, `SdcaL1UpdateSparse`, and `ZeroMatrixItems` (might exist in System.Memory).
- Generic overloads of all relevant methods, constrained to the appropriate numerical types

Concrete proposal to follow.

**Comment by stephentoub** [comment]
> Could you please elaborate on the advantages of having these APIs in a BCL rather than in a specialized NuGet package

It _is_ a nuget package today. It's currently not part of netcoreapp. If it were to be pulled into netcoreapp, it would be because we'd be using it from elsewhere in netcoreapp, e.g. using it from APIs like Enumerable.Average, BitArray.And, WebSocket, etc.

**Comment by xoofx** [comment]
Hey @stephentoub,

Would it be possible to expose the low level parts of the API instead of only providing Span versions?

e.g 

```c#
public static Vector128<float> Log2(Vector128<float> value);
public static Vector256<float> Log2(Vector256<float> value);
public static Vector512<float> Log2(Vector512<float> value);
//...etc.
```

I did that for a prototype for a similar API and it's working great. 
One reason to expose these APIs is that you can actually build higher level functions (e.g for tensors, the zoo of the activation functions) and build Span versions on top of them.

These API can then be used for other kind of custom Span batching (not related to tensors), where the packing of the vector is different (e.g 4xfloat chuncked `xxxx, yyyy, zzzz`)

**Comment by tannergooding** [comment]
> Would it be possible to expose the low level parts of the API instead of only providing Span versions?

Yes, but it needs to be its own proposal and cover all 5 vector types (Vector<T>, Vector64/128/256/512<T>) and consider whether its applicable to Vector2/3/4 as well.

**Comment by xoofx** [comment]
> Yes, but it needs to be its own proposal and cover all 5 vector types (Vector, Vector64/128/256/512)

Cool, I will try to write something.

**Comment by xoofx** [comment]
> Would it be possible to expose the low level parts of the API instead of only providing Span versions?

Follow-up, created the proposal #93513

**Comment by msedi** [comment]
@stephentoub: 

> If it were to be pulled into netcoreapp as well, it would be because we'd be using it from elsewhere in netcoreapp

if brought to the BCL wouldn't it make sense to rename `TensorPrimitives` to lets say `ArrayMath`, `VectorMath` or `VectorPrimitives`. Tensor seems a bit exaggerated for what it does, namely doing some math on arrays.

**Comment by tannergooding** [comment]
@msedi that would be a breaking change. Additionally, the intent is to expand it to the full set of BLAS support, so Tensor is a very apt and appropriate name that was already scrutinized, reviewed, and approved by API review

**Comment by msedi** [comment]
@tannergooding: Sure you right, I was just under the impression that there could be something more primitive. The tensor ist something, lets say higher level whereas the vector/array methods are on a lower level. But I'm completely fine with it whenever I know where to find it,

BTW. When looking at the code and the effort for the TensorPrimitives are there any efforts the JIT will some day manage to do the SIMD unfolding for us?

**Comment by tannergooding** [comment]
> the JIT will some day manage to do the SIMD unfolding for us?

The JIT is unlikely to get auto-vectorization in the near future as such support is complex and quite expensive to do. Additionally, outside of particular domains, such support does not often light up and has measurable impact to real world apps even less frequently. Especially for small workloads it can often have the opposite effect and slow down your code. In the domains where it does light up, and particularly where it would be beneficial to do, you are often going to get better perf by writing your own SIMD code directly.

It is therefore my opinion that our efforts would be better spent providing APIs from the BCL that provide this acceleration for you. Such as all the APIs on `Span<T>`, accelerating LINQ, the new APIs on `TensorPrimitives`, etc. It may likewise be beneficial to expose some SIMD infrastructure helpers like we've defined internally for `TensorPrimitives`; that is expose some public form of `InvokeSpanSpanIntoSpan` and friends, which would allow developers to only worry about providing the inner kernel and to have the rest of the SIMD logic (leading/trailing elements, alignment, unrolling, etc) handled internally. Efforts like `ISimdVector<TSelf, T>` also fit the bill of making it simpler for devs to write SIMD code.

**Comment by msedi** [comment]
@tannergooding : Thanks for the info. That makes since. For our case we wrote source generators to do generate all the array primitives, currently with Vector<T> but I wanted to benchmark against your implementations. I assume yours is better ;-)

**Comment by tannergooding** [comment]
Remaining work is for .NET 10

**Comment by ds5678** [comment]
`BitwiseXor` seems to be marked above as implemented, but it doesn't show up in the code search.

https://github.com/search?q=repo%3Adotnet%2Fruntime+BitwiseXor&type=code

**Comment by stephentoub** [comment]
~~BitwiseXor wasn't in https://github.com/dotnet/runtime/issues/96451, which is why it's not in the source. @tannergooding, do you remember if there was a reason we didn't include it?~~

**Comment by stephentoub** [comment]
Oh, wait, no, it is there, we just called it Xor rather than BitwiseXor. I know we discussed the naming but I don't remember why we chose to name this one differently.

https://github.com/dotnet/runtime/blob/83ffb89afaf1b3eb0a2557fde76faffb00f1ef0f/src/libraries/System.Numerics.Tensors/ref/System.Numerics.Tensors.netcore.cs#L788-L789

**Comment by tannergooding** [comment]
Because it matches the official alternative names: https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2225

I personally would prefer if they were just And, Or, and xor because it better matches the flow for names like AndNot or Xnor and other common combinations. But it's an existing convention

**Comment by jeffhandley** [comment]
Carrying this epic forward into .NET 11 for the items that remain open.
